sdk-ext-ui/src/androidTest/java/cash/z/ecc/sdk/ext/ui/regex/ZecRegexTest.kt

```diff
@@ -0,0 +1,86 @@
+package cash.z.ecc.sdk.ext.ui.regex
+
+import androidx.test.filters.SmallTest
+import cash.z.ecc.sdk.ext.ui.R
+import cash.z.ecc.sdk.ext.ui.test.getStringResourceWithArgs
+import cash.z.ecc.sdk.model.MonetarySeparators
+import org.junit.Test
+import kotlin.test.assertFalse
+import kotlin.test.assertNotNull
+import kotlin.test.assertTrue
+
+class ZecRegexTest {
+
+    @Test
+    @SmallTest
+    fun check_regex_validity() {
+        val separators = MonetarySeparators.current()
+        val regexString = getStringResourceWithArgs(
+            R.string.zec_amount_regex_filter,
+            arrayOf(
+                separators.grouping,
+                separators.decimal
+            )
+        )
+        assertNotNull(regexString)
+
+        val regexAmountChecker = regexString.toRegex()
+
+        regexAmountChecker.also {
+            assertNotNull(regexAmountChecker)
+            assertTrue(regexAmountChecker.pattern.isNotEmpty())
+        }
+    }
+
+    @Test
+    @SmallTest
+    fun check_regex_functionality_valid_inputs() {
+        val separators = MonetarySeparators.current()
+        val regex = getStringResourceWithArgs(
+            R.string.zec_amount_regex_filter,
+            arrayOf(
+                separators.grouping,
+                separators.decimal
+            )
+        ).toRegex()
+
+        regex.also {
+            assertTrue(it.matches(""))
+            assertTrue(it.matches("123"))
+            assertTrue(it.matches("${separators.decimal}"))
```

Comment:
I think we are going to have trouble with some, but not all, of these test cases in the future.  The behavior of the test case depends on the locale of the device it is running on (and the availability of localized resources for RTL).  To make this easier on our future selves, a simple solution might be:

1. Add annotation to require API 17 for this test case.  I think it is `@SdkSuppress(minSdkVersion = Build.VERSION_CODES.N)`  We haven't settled on min API yet, but 17 is reasonable and required for the next suggestion.  (We're currently targeting 16, hence the need for the annotation)

2. Use an if statement checking for [LTR](https://developer.android.com/reference/android/text/TextUtils.html#getLayoutDirectionFromLocale(java.util.Locale) ) for assertions that would matter on direction.  (That means primarily the ones that have both separators and numbers)

3. Don't worry about adding RTL test cases in the else block.  Instead add a TODO pointing to https://github.com/zcash/secant-android-wallet/issues/248

This comment would apply to both here and in the Compose UI test.

Reply:
Thank you @ccjernigan for the proposed solutions. I think that the solution **2 - tests for both LTR and RTL** is the way we should go. I've just tested how the OS treats our test with **Force RTL** developer option as well as with switch device language to **Arabic** on both real and emulator devices, and I believe that we don't nee to make any changes in our regex patter in strings.xml nor in implementation of our tests. Even if the method `TextUtils.getLayoutDirectionFromLocale(locale)` returns us RTL, the regex and tests for LTR succeed. It seems that RTL vs LTR only affects UI displaying and our tests are still valid.

What do you think about it?

Reply:
I found another problem in Arabic locale and that it related to `ZecString.fromZecString` method. 
`IllegalArgumentException: Unquoted special character '#' in pattern "#٬##0٫0#"`
So I've added TODO pointing to https://github.com/zcash/secant-android-wallet/issues/248. You probably also noticed as you have it marked in the pointing issue.

Reply:
I want to make sure we're on the same page:

The test cases won't fail right now because we're missing the RTL localization in strings.xml.  But once it is added, the all the test cases are broken.  E.g. Testing against a hard-coded string of "1,234.00" breaks once the RTL strings.xml is added since the order should be reversed.  I am not suggesting we add a RTL localization or fix those test cases now.

I'm only suggesting adding the if statement in the tests so that when we come back to this both A. the tests won't fail immediately and B. the hook for adding additional test cases is present.  I am suggesting this because code has a habit of being copy-pasted, and I don't want hard-coded localization issues in our tests to propagate to other screens.  This makes others who review the code aware that we know RTL is not fully supported here.

Reply:
That's a great catch in fromZecString.

Does it crash when the system language is set to Arabic and then we try to send ZEC?  That sounds like a bug, because someone with an RTL language should basically see the app work as it does in English until we add the right RTL resources.  If that is true, then it might need to be a separate bug report because #248 is about adding a new feature (supporting RTL in the UI).

Reply:
Discussed more in details out of this thread. For now we suppose that the tests won't be affected once we add RTL language to the app in the future.

About the crash on the `ZEC Request` screen after typing a single character to amount field - filed a new issue https://github.com/zcash/secant-android-wallet/issues/321 and linked form implementation too.